src/mono/mono/mini/debugger-protocol.c

```diff
@@ -1,8 +1,18 @@
 #include "debugger-protocol.h"
+
+#ifdef DBI_COMPONENT
+#define g_malloc malloc
+#define g_free free
+#define g_assert assert
+#define g_realloc realloc
+#define mono_atomic_inc_i32 InterlockedIncrement
+#include "stdafx.h"
+#else
+#include <glib.h>
 #include <mono/utils/atomic.h>
-#include "glib.h"
+#endif
 
-static int packet_id = 0;
+static LONG packet_id = 0;
```

Comment:
All of this feels pretty fragile and a bit hacky.

But the piece I like the least is `LONG` and that define for atomic increment.
What if you added
```c
static inline int32_t
atomic_inc_i32 (volatile int32_t *dest)
{
	return InterlockedIncrement ((LONG volatile *)dest);
}
```

And then defined `static int32_t packet_id;`

Reply:
It might be better to add an explicit header that defines your interop interface, and then 2 implementation headers - one for mono using glib, and one for the dbi component using something compatible with the cordb.

Reply:
Similar was done in EventPipe port (but on much bigger scale), For example, https://github.com/dotnet/runtime/blob/master/src/native/eventpipe/ep-rt.h, https://github.com/dotnet/runtime/blob/master/src/native/eventpipe/ep-rt-types.h and  defines the API (one of them) implemented by all runtime that would like to consume the library and then each runtime implements them mostly using inlineable functions, here you find mono's implementation, https://github.com/dotnet/runtime/blob/master/src/mono/mono/eventpipe/ep-rt-mono.h, https://github.com/dotnet/runtime/blob/master/src/mono/mono/eventpipe/ep-rt-types-mono.h. The whole EventPipe library standardized on C99 data types and is consumed by both C and C++ clients. For the atomic functions you can find similar abstractions in these API's, so this is what you find in the shim API (ep-rt.h in this case):

```
static
int32_t
ep_rt_atomic_inc_int32_t (volatile int32_t *value);
```

and this is what gets used when building library for Mono:

```
static
inline
int32_t
ep_rt_atomic_inc_int32_t (volatile int32_t *value)
{
	return (int32_t)mono_atomic_inc_i32 ((volatile gint32 *)value);
}
```

and on CoreClr:

```
static
inline
int32_t
ep_rt_atomic_inc_int32_t (volatile int32_t *value)
{
	STATIC_CONTRACT_NOTHROW;
	return static_cast<int32_t>(InterlockedIncrement ((volatile LONG *)(value)));
}
```

source needs of course to adapt to use the shim API instead of g_* symbols, but that will make the implementation more straight forward to port over to other runtimes.

Reply:
Thanks! Done!